### ISSUE TITLE
Support batched control sequences in `dsx.simulate`

### ISSUE THREAD

**Issue by MatthieuDarcy** [BasisResearch/dynestyx#347]
`dsx.simulate(..., n_simulations=N)` currently draws `N` trajectories that differ only
in their initial condition. 

Sampling based MPC (such as MPPI) would like to batch over control sequences. 

### Tentative recommendation 

Allow `ctrl_values` to carry an optional leading simulation axis:

```python
ctrl_values: Real[Array, "ctrl_time control_dim"]          # shared (current behavior)
           | Real[Array, "n_controls ctrl_time control_dim"]  # per-trajectory
```
We would have to think about how this interact with `n_simulations`.
1. Option 1: creates $N \times n_c$ simulation (each control has $N$ simulations attached). Could be useful if we want to have the same $N$ initial conditions/randomness, each paired with a seperate control. This is the most useful in MPPI.
2. Option 2: assume that $N= n_c$, sample initial conditions/randomness seperately per control. Note that a user could duplicate controls to achieve $N\times n_c$ trajectories (with the caveat that randomness is not shared in this case).

**Comment by mattlevine22** [comment]
Can the user simply do `vmap` over the controls?

**Comment by MatthieuDarcy** [comment]
Do you mean something like
```

def run_single(control):
    res = dsx.simulate(..., control)
    return res

res_batch = vmap(run_single)(control_batch)
```

If so, the answer is no. SimulatedResult isn't currently registered with jax.tree_util, so it can't cross a vmap boundary. 

So what I do in MPPI is:

```
def compute_loss(control):
    res = dsx.simulate(...)
    return loss(res), res.states, res.observations

losses, states, observations = vmap(compute_loss)(control_batch)
```

This works because `res` is consumed inside the vmapped function and only plain arrays come out. But I think it would be useful to have access to the SimulatedResult directly in some cases and would avoid having to extract states and observations directly.

If I'm missing something and there is a better way to do this, I would love to hear it, it would considerably simplify my MPPI implementation!

**Comment by mattlevine22** [comment]
Digging a bit deeper, it seems like the reason `SimulatedResult` cannot be registered w/ `jax.tree_util` is the annoying `_register_numpyro_sites` field. Fortunately, you don't need that at all here (it is only used in PPL mode so-to-speak).

Codex suggests:

```python
import dataclasses
import jax


@jax.tree_util.register_dataclass
@dataclasses.dataclass
class SimulatedResult:
    times: Array | None = None
    x_0: Array | None = None
    states: Array | None = None
    observations: Array | None = None
    predicted_times: Array | None = None
    predicted_states: Array | None = None
    predicted_observations: Array | None = None

    _register_numpyro_sites: Callable[[str], None] | None = dataclasses.field(
        default=None,
        repr=False,
        metadata={"static": True},
    )
```

**Comment by MatthieuDarcy** [comment]
Sounds good, will take a look

**Comment by DanWaxman** [comment]
> Digging a bit deeper, it seems like the reason `SimulatedResult` cannot be registered w/ `jax.tree_util` is the annoying `_register_numpyro_sites` field. Fortunately, you don't need that at all here (it is only used in PPL mode so-to-speak).
> 
> Codex suggests:
> 
> import dataclasses
> import jax
> 
> 
> @jax.tree_util.register_dataclass
> @dataclasses.dataclass
> class SimulatedResult:
>     times: Array | None = None
>     x_0: Array | None = None
>     states: Array | None = None
>     observations: Array | None = None
>     predicted_times: Array | None = None
>     predicted_states: Array | None = None
>     predicted_observations: Array | None = None
> 
>     _register_numpyro_sites: Callable[[str], None] | None = dataclasses.field(
>         default=None,
>         repr=False,
>         metadata={"static": True},
>     )

I think this is basically what an `equinox.Module` is, I would try to use that instead. Indeed, things seem to work in #350 -- the following test passes

```python

def test_simulate_vmaps_over_batched_controls():
    dynamics = _make_controlled_deterministic_discrete_dynamics()
    predict_times = jnp.arange(4.0)
    control_batch = jnp.array(
        [
            [[1.0], [1.0], [1.0], [1.0]],
            [[2.0], [2.0], [2.0], [2.0]],
        ]
    )

    result = jax.vmap(
        lambda ctrl_values: dsx.simulate(
            dynamics,
            rng_key=jr.PRNGKey(0),
            ctrl_times=predict_times,
            ctrl_values=ctrl_values,
            predict_times=predict_times,
            n_simulations=2,
        )
    )(control_batch)

    assert isinstance(result, dsx.SimulatedResult)
    assert result.states is not None
    assert result.states.shape == (2, 2, 4, 1)
    expected_states = jnp.array(
        [
            [[[0.0], [1.0], [2.0], [3.0]]],
            [[[0.0], [2.0], [4.0], [6.0]]],
        ]
    )
    assert jnp.array_equal(
        result.states, jnp.broadcast_to(expected_states, (2, 2, 4, 1))
    )
```

Let me know if that's no good for some reason.
